jlm/llvm/backend/dot/DotWriter.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/llvm/backend/dot/DotWriter.hpp>
+
+#include <jlm/llvm/ir/types.hpp>
+#include <jlm/rvsdg/bitstring/type.hpp>
+#include <jlm/rvsdg/region.hpp>
+#include <jlm/rvsdg/simple-node.hpp>
+#include <jlm/rvsdg/structural-node.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/rvsdg/type.hpp>
+
+namespace jlm::llvm::dot
+{
+/**
+ * Creates a node in the \p typeGraph representing the given \p type,
+ * or returns such a node if it has already been created.
+ * The function is recursive, and will create nodes for subtypes of aggregate types.
+ */
+static util::Node &
+GetOrCreateTypeGraphNode(const rvsdg::type & type, util::Graph & typeGraph)
+{
+  // If the type already has a corresponding node, return it
+  if (auto * graphElement = typeGraph.GetElementFromProgramObject(type))
+  {
+    auto * node = reinterpret_cast<util::Node *>(graphElement);
+    JLM_ASSERT(node);
+    return *node;
+  }
+
+  auto & node = typeGraph.CreateNode();
+  node.SetProgramObject(type);
+  node.SetLabel(type.debug_string());
+
+  // Some types get special handling, such as adding incoming edges from aggregate types
+  if (jlm::rvsdg::is<rvsdg::statetype>(type) || jlm::rvsdg::is<rvsdg::bittype>(type)
+      || jlm::rvsdg::is<PointerType>(type) || jlm::rvsdg::is<fptype>(type)
+      || jlm::rvsdg::is<varargtype>(type))
```

Comment:
More like FYI, but it should not be necessary to prefix everything with `jlm::`. `rvsdg::` should be sufficient.

Reply:
If I am mistaken here, then it might be beneficial to use `using namespace jlm::rvsdg`.

Reply:
You are right. For some reason CLion did not gray out the `jlm::` part.